Replace current regex engine with PCRE2

Rot127 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Replace the current regex engine with PCRE2.

**Test plan**

All green.

**Closing issues**

closes https://github.com/rizinorg/rizin/issues/3730

Partially addresses https://github.com/rizinorg/rizin/issues/4055

**Todo**

- [x] Add a `rz_vector_pop_ptr()` function which doesn't use `memcpy()` and use it in `match_all_flat()`. Or add a `rz_vector_concat()` function with the same functionality.
- [x] Terminology for groups and matches is weird.
- [x] docs
- [ ] Open bug report for tcc and pcre2 compilation and `__asm`
- [x] Test `rz_regex_get_match_name`

Rot127 committed:
Replace OpenBSD regex library with PCRE2.

PCRE2 has way better performance than the OpenBSD
library (something around 20 times faster).

The following flags are enabled for every pattern:

- PCRE2_UTF
- PCRE2_MATCH_INVALID_UTF
- PCRE2_NO_UTF_CHECK

All the others are optional.

Changes made:

- Adds PCRE2 as subproject.
- Changes the API away from POSIX to PCRE2.
- Edits many regex patterns because:
 - ' ' is skipped in patterns, if the EXTENDED flag is set for matching. '\s' must be set now.
 - '.' doesn't match newlines by default.
- Changes the API so matches and their groups are bundled into PVectors.
- Moves the regex component to rz_util.

Rot127 committed:
Test for POSIX blank.

XVilka reviewed (approved):
LGTM if all green

XVilka reviewed (changes requested):
Cross-build fails:
```
librz/util/meson.build:172:19: ERROR: Tried to mix libraries for machines build machine and host machine in target 'rz_util_native' This is not possible in a cross build.
```
 Just define native dependency for cross-build separately, like other libs in the same file

Rot127 committed:
Fix cross build - add copy of PCRE2 dependecy

Meson currently doesn't support subprojects to be native and non-native at the same time.
See: https://github.com/mesonbuild/meson/issues/10947
Unfortunately, sdb depends on rz_util which in turn depends on PCRE2.
Excluding PCRE2 from the native build makes linking of rz_util not possible anymore.
Adding it, will make Meson complain that the dependencies cannot be mixed.

Hence, we compile a copy of PCRE2 for the native build if required.

wargio commented on librz/cons/less.c:
```diff
@@ -3,9 +3,10 @@
 
 #include <rz_cons.h>
-#include <rz_regex.h>
+#include <rz_util/rz_regex.h>
 #include <rz_util.h>
```
`#include <rz_util/rz_regex.h>` should be added in `#include <rz_util.h>`

wargio commented on librz/cons/less.c:
```diff
@@ -3,9 +3,10 @@
+#include <rz_util/rz_regex.h>
 #include <rz_util.h>
 #include "pager_private.h"
+#include "rz_vector.h"
```
also this is global not local.

arrowd commented:
Why use static linking here? I'm getting
```
ld: error: relocation R_X86_64_32S cannot be used against symbol '_pcre2_ucd_records_8'; recompile with -fPIC
```
on FreeBSD when building against system pcre2.

wargio commented:
@Rot127

Rot127 commented:
Thanks for the report! Will fix it in two days.

If you need a quick fix, you can add the C flag in  `subprojects/packagefiles/pcre2/meson.build`.
And run `git clean -dxff subprojects/pcre2*` afterwards.
